src/components/Search/Search.js

```diff
@@ -26,25 +65,157 @@ const Search = React.createClass({
+          newIndex = -1;
+        }
+        break;
+      case "ArrowUp":
+        if (index > -1) {
+          newIndex--;
+          if (index < numSuggestions) {
+            // We are in suggestions, move on up.
+            newSuggestionIndex--;
+          } else if (index === numSuggestions) {
+            // We are on the first engine, reset engine index and move to
+            // last suggestion.
+            newEngineIndex = -1;
+            newSuggestionIndex = numSuggestions - 1;
+          } else if (index < numSuggestions + numEngines) {
+            // We are on the engine list, move on up.
+            newEngineIndex--;
+          } else {
+            // We are on the button, move to last engine.
+            newEngineIndex = numEngines - 1;
+          }
+        } else {
+          // Nothing is selected, go to the very end.
+          newIndex = numSuggestions + numEngines;
+        }
+        break;
+      case "Tab":
+        // Tab only navigates through the engines list.
+        if (!evt.shiftKey) {
+          // Shift isn't pressed, go forward.
+          if (index === numSuggestions + numEngines) {
+            // We reached the end, let the event go on.
+            return;
+          }
+          if (index < numSuggestions) {
+            // We aren't in the engines list yet, move to first engine.
+            newIndex = numSuggestions;
+            newEngineIndex = 0;
+          } else {
+            // We are in the engines list, move along.
+            newIndex++;
+            newEngineIndex++;
+          }
+        } else {
+          // Shift is pressed, go backward.
+          if (index < numSuggestions) {
+            // We aren't on the engines list, ;et the event move on.
+            return;
+          }
+          if (index === numSuggestions) {
+            // We are on the first engine, unselect it and go to where we were
+            // in the suggestions list.
+            newEngineIndex = -1;
+            newIndex = newSuggestionIndex;
+          } else if (index < numSuggestions + numEngines) {
+            // We are in the engines list, move up the list.
+            newIndex--;
+            newEngineIndex--;
+          } else {
+            // We are on the button, go to bottom of engine list.
+            newIndex--;
+            newEngineIndex = numEngines - 1;
+          }
+        }
+        break;
+      case "Enter":
+        evt.preventDefault();
+        // If the change settings button is selected, fire the action for it.
+        if (this.getSettingsButtonIsActive()) {
+          Platform.search.manageEngines();
+          return;
+        }
+
+        // Otherwise, perform the search with active engine and suggestion.
+        this.performSearch({
+          engineName: this.getActiveEngine(),
+          searchString: this.getActiveSuggestion() || this.props.Search.searchString
+        });
+        return;
+      default:
+        return;
+    }
+
+    evt.preventDefault();
```

Comment:
Ah, ok, it might be in case of tab.... maybe reorder the case statement so "tab" is before default? Also, maybe add a note here that this code only applies to "tab"... or maybe move this code?